2024-04-11 - CI failed in ruby-head(3.4.0dev)

https://github.com/kachick/irb-power_assert/actions/runs/8640409928/job/23688270555

```
  /home/runner/.rubies/ruby-head/bin/ruby --version
  ruby 3.4.0dev (2024-04-10T17:33:44Z :detached: d75dc39880) [x86_64-linux]
  Took   0.03 seconds
```

```
Resolving dependencies...
Writing lockfile to /home/runner/work/irb-power_assert/irb-power_assert/Gemfile.lock
Cache key: setup-ruby-bundler-cache-v6-ubuntu-22.04-x64-ruby-head-wd-/home/runner/work/irb-power_assert/irb-power_assert-with--without-development-only--ABI-3.4.0+0-Gemfile.lock-33a477bc319d5d34d7c48c33a798344746ded50a269cb0854f6c51680331f8c4
Cache Size: ~1 MB (584916 B)
/usr/bin/tar -xf /home/runner/work/_temp/9db84b9e-2ca4-4bcb-8fec-88f87df4fa35/cache.tzst -P -C /home/runner/work/irb-power_assert/irb-power_assert --use-compress-program unzstd
Cache restored successfully
Found cache for key: setup-ruby-bundler-cache-v6-ubuntu-22.04-x64-ruby-head-wd-/home/runner/work/irb-power_assert/irb-power_assert-with--without-development-only--ABI-3.4.0+0-Gemfile.lock-33a477bc319d5d34d7c48c33a798344746ded50a269cb0854f6c51680331f8c4
/home/runner/.rubies/ruby-head/bin/bundle install --jobs 4
Bundle complete! 7 Gemfile dependencies, 12 gems now installed.
Gems in the group 'development' were not installed.
Bundled gems are installed into `./vendor/bundle`
Took   2.[30](https://github.com/kachick/irb-power_assert/actions/runs/8640409928/job/23688270555#step:3:36) seconds
```

```
/home/runner/work/irb-power_assert/irb-power_assert/vendor/bundle/ruby/3.4.0+0/gems/rake-13.1.0/lib/rake.rb:33: warning: ostruct was loaded from the standard library, but will no longer be part of the default gems since Ruby 3.5.0. Add ostruct to your Gemfile or gemspec.
can not use rubocop in this environment
===============================================================================
Error: test_typical_example(TestIRBPowerAssert): NoMethodError: undefined method 'irb_pa' for #<TestIRBPowerAssert:0x0000[7](https://github.com/kachick/irb-power_assert/actions/runs/8640409928/job/23688270555#step:4:8)fc9bb6c9210>
/home/runner/work/irb-power_assert/irb-power_assert/test/test_irb_power_assert.rb:44:in 'block in TestIRBPowerAssert#test_typical_example'
/home/runner/work/irb-power_assert/irb-power_assert/test/helper.rb:37:in 'Test::Unit::TestCase#capture_io'
/home/runner/work/irb-power_assert/irb-power_assert/test/test_irb_power_assert.rb:43:in 'TestIRBPowerAssert#test_typical_example'
===============================================================================
Finished in 0.007[8](https://github.com/kachick/irb-power_assert/actions/runs/8640409928/job/23688270555#step:4:9)4925 seconds.
2 tests, 3 assertions, 0 failures, 1 errors, 0 pendings, 0 omissions, 0 notifications
rake aborted!
Command failed with status (1)
<internal:array>:52:in 'Array#each'
<internal:array>:52:in 'Array#each'
/home/runner/work/irb-power_assert/irb-power_assert/vendor/bundle/ruby/3.4.0+0/gems/rake-13.1.0/exe/rake:27:in '<top (required)>'
/home/runner/.rubies/ruby-head/bin/bundle:25:in 'Kernel#load'
/home/runner/.rubies/ruby-head/bin/bundle:25:in '<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
Error: Process completed with exit code 1.
```

--- a/test/helper.rb
+++ b/test/helper.rb
@@ -25,22 +25,46 @@ end
 
 require_relative '../lib/irb-power_assert'
 
-class Test::Unit::TestCase
-  # Taken from https://github.com/ruby/irb/blob/ad08152c43d4309ee4dec3bbaf361ffc338c1f46/test/lib/minitest/unit.rb#L461-L495, thank you!
-  def capture_io
-    captured_stdout, captured_stderr = StringIO.new, StringIO.new
+module TestIRBPowerAssertHelpers
+  class TestInputMethod < ::IRB::InputMethod
+    attr_reader :list, :line_no
 
-    orig_stdout, orig_stderr = $stdout, $stderr
-    $stdout, $stderr         = captured_stdout, captured_stderr
-
-    begin
-      yield
-    ensure
-      $stdout = orig_stdout
-      $stderr = orig_stderr
+    def initialize(list = [])
+      super()
+      @line_no = 0
+      @list = list
     end
 
-    return captured_stdout.string, captured_stderr.string
+    def gets
+      @list[@line_no]&.tap { @line_no += 1 }
+    end
+
+    def eof?
+      @line_no >= @list.size
+    end
+
+    def encoding
+      Encoding.default_external
+    end
+
+    def reset
+      @line_no = 0
+    end
   end
-  alias_method :capture_output, :capture_io
+
+  def execute_lines(*lines, conf: {}, main: self, irb_path: nil)
+    raise unless RUBY_VERSION >= '3.4'
+
+    IRB.init_config(nil)
+    IRB.conf[:VERBOSE] = false
+    IRB.conf[:PROMPT_MODE] = :SIMPLE
+    IRB.conf[:USE_PAGER] = false
+    IRB.conf.merge!(conf)
+    input = TestInputMethod.new(lines)
+    irb = IRB::Irb.new(IRB::WorkSpace.new(main), input)
+    irb.context.return_format = "=> %s\n"
+    irb.context.irb_path = irb_path if irb_path
+    IRB.conf[:MAIN_CONTEXT] = irb.context
+    irb.eval_input
+  end
 end

--- a/test/test_irb_power_assert.rb
+++ b/test/test_irb_power_assert.rb
@@ -6,6 +6,8 @@ require_relative 'helper'
 require 'stringio'
 
 class TestIRBPowerAssert < Test::Unit::TestCase
+  include TestIRBPowerAssertHelpers
+
   def test_constants
     assert(IRB::PowerAssert::VERSION.frozen?)
     assert do
@@ -14,19 +16,11 @@ class TestIRBPowerAssert < Test::Unit::TestCase
 
     am = method(:assert_match)
     ::PowerAssert.class_exec do
-      am.call(%r{/irb-power_assert/lib\z}, const_get(:INTERNAL_LIB_DIRS)[IRB::PowerAssert])
+      am.call(%r{#{Regexp.escape(Dir.pwd)}/lib}, const_get(:INTERNAL_LIB_DIRS)[IRB::PowerAssert])
     end
   end
 
   def test_typical_example
-    IRB.setup(__FILE__, argv: [])
-    IRB.conf[:USE_MULTILINE] = true
-    IRB.conf[:USE_SINGLELINE] = false
-    IRB.conf[:VERBOSE] = false
-    workspace = IRB::WorkSpace.new(self)
-    irb = IRB::Irb.new(workspace)
-    IRB.conf[:MAIN_CONTEXT] = irb.context
-
     expected =<<~'EOD'
     result: false
 
@@ -41,9 +35,24 @@ class TestIRBPowerAssert < Test::Unit::TestCase
     EOD
 
     out, err = capture_output do
-      irb.context.main.irb_pa(%q{"0".class == "3".to_i.times.map {|i| i + 1 }.class})
+      # We can't handle with the IRB version, because they do not bump version number in development versions.
+      if RUBY_VERSION > '3.4'
+        # Since ruby 3.4 (or newer irb gem), no need irbrc hack to realize `pa exp` instead of `pa 'exp'`
+        execute_lines(%q{pa "0".class == "3".to_i.times.map {|i| i + 1 }.class})
+      else
+        IRB.setup(__FILE__, argv: [])
+        IRB.conf[:USE_MULTILINE] = true
+        IRB.conf[:USE_SINGLELINE] = false
+        IRB.conf[:VERBOSE] = false
+        workspace = IRB::WorkSpace.new(self)
+        irb = IRB::Irb.new(workspace)
+        IRB.conf[:MAIN_CONTEXT] = irb.context
+
+        irb.context.main.irb_pa(%q{"0".class == "3".to_i.times.map {|i| i + 1 }.class})
+      end
     end
+
     assert_equal('', err)
-    assert_equal(expected, out)
+    assert_equal((RUBY_VERSION > '3.4' ? expected + "=> nil\n" : expected), out)
   end
 end

